exonum-java-binding-common/src/main/java/com/exonum/binding/common/message/TransactionMessage.java

```diff
@@ -0,0 +1,174 @@
+import com.exonum.binding.common.crypto.PublicKey;
+import com.exonum.binding.common.hash.HashCode;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * An Exonum transaction message.
+ */
+public interface TransactionMessage {
+  int AUTHOR_PUBLIC_KEY_OFFSET = 0;
+  int CLS_OFFSET = 32;
+  int TAG_OFFSET = 33;
+  int SERVICE_ID_OFFSET = 34;
+  int TRANSACTION_ID_OFFSET = 36;
+  int PAYLOAD_OFFSET = 38;
+
+  int AUTHOR_PUBLIC_KEY_SIZE = 32;
+  int SIGNATURE_SIZE = 64;
+
+  /**
+   * Returns a public key of the Transaction Message's author.
+   */
+  PublicKey getAuthor();
+
+  /**
+   * Returns service id of the Transaction Message.
+   */
+  short getServiceId();
+
+  /**
+   * Returns transaction id of the Transaction Message.
+   */
+  short getTransactionId();
+
+  /**
+   * Returns payload of the Transaction Message.
+   */
+  byte[] getPayload();
+
+  /**
+   * Returns Transaction Message hash.
+   */
+  HashCode hash();
+
+  /**
+   * Returns Transaction Message signature.
+   */
+  byte[] getSignature();
+
+  /**
+   * Returns Transaction Message in binary format.
+   */
+  byte[] toBytes();
+
+  /**
+   * Returns the Builder for the Transaction Message.
+   */
+  static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Creates Transaction Message from the given bytes array.
+   */
+  static TransactionMessage fromBytes(byte[] bytes) {
+    return BinaryTransactionMessage.fromBuffer(ByteBuffer.wrap(bytes));
+  }
+
+  /**
+   * Creates Transaction Message from the given bytes buffer.
+   */
+  static TransactionMessage fromBuffer(ByteBuffer buffer) {
+    return BinaryTransactionMessage.fromBuffer(buffer);
+  }
+
+  /**
+   * Transaction Message Builder class.
+   */
+  class Builder {
+    private Short serviceId;
+    private Short transactionId;
+    private ByteBuffer payload;
+
+    /**
+     * Sets service id to the transaction message.
+     */
+    public Builder serviceId(short serviceId) {
+      this.serviceId = serviceId;
+      return this;
+    }
+
+    /**
+     * Sets transaction id to the transaction message.
+     */
+    public Builder transactionId(short transactionId) {
+      this.transactionId = transactionId;
+      return this;
+    }
+
+    /**
+     * Sets payload to the transaction message.
+     */
+    public Builder payload(byte[] payload) {
+      return payload(ByteBuffer.wrap(payload));
+    }
+
+    /**
+     * Sets payload to the transaction message.
+     */
+    public Builder payload(ByteBuffer payload) {
+      this.payload = payload.duplicate().order(ByteOrder.LITTLE_ENDIAN);
+      return this;
+    }
+
+    /**
+     * Signs and creates an instance of the transaction message.
+     *
+     * @throws NullPointerException if serviceId or transactionId or payload haven't set
+     * @throws IllegalArgumentException if public key has wrong size
+     */
+    public TransactionMessage sign(KeyPair keys, CryptoFunction crypto) {
+      checkNotNull(serviceId);
```

Comment:
Something like `checkRequiredFieldSet(field: Object, fieldName: String)` so that clients do not get plain NPEs, but, say, IllegalStateException with message that explains things?